### ISSUE TITLE
regression: PCH with msvc causing "error LNK2011: precompiled object not linked in; image may not run"

### ISSUE THREAD

**Issue by anarazel** [mesonbuild/meson#10745]
Hi,

As mentioned on IRC: Starting with @dcbaker 's c94c492089e7fecb56b7cc11ea76712a770f1e34 I see msvc builds of postgres fail when pch is enabled:
```
[56/56] Linking target src/bin/pg_config/pg_config.exe
FAILED: src/bin/pg_config/pg_config.exe src/bin/pg_config/pg_config.pdb
"link"  /MACHINE:x64 /OUT:src/bin/pg_config/pg_config.exe src/bin/pg_config/pg_config.exe.p/pg_config.c.obj "/nologo" "/release" "/nologo" "/DEBUG" "/PDB:src\bin\pg_config\pg_config.pdb" "src/common\libpgcommon.a" "src/fe_utils\libpgfeutils.a" "src/port\libpgport.a" "ws2_32.lib" "ws2_32.lib" "ws2_32.lib" "ws2_32.lib" "/SUBSYSTEM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
libpgcommon.a(config_info.c.obj) : error LNK2011: precompiled object not linked in; image may not run
  Hint on symbols that are defined and could potentially match:
    __@@_PchSym_@00@UwveUklhgtivhRnvhlmUyfrowRmrmqzUhixUkligUoryktkligOzOkUxPkxsOlyq@pgport_main
    __@@_PchSym_@00@UwveUklhgtivhRnvhlmUyfrowRmrmqzUhixUxlnnlmUoryktxlnnlmOzOkUxPkxsOlyq@pgcommon
```

The @pgport_main etc is the result of me going through the static libraries and naming them with `/Yl<somename>` as part of trying to debug this.

I spent a decent bit of time trying to build a reproducer, but I'm getting too tired.  I suspect it's related to a static library that links to another static library. But it doesn't trigger in a simplistic reproducer unfortunately.

Regards,

Andres

**Comment by anarazel** [comment]
Just when I was ready to give up I reproduced it. For some reason the example needs link_whole, while postgres's case doesn't need link_whole. Assume somehow postgres's link_with gets transformed to a link_whole.

https://github.com/anarazel/meson/commit/90bfa3fdf4b6cedef9683d56c9ec3d4bb8b8a091 hacks the existing test up to show the problem. There's lots of gunk in there that needs to be cleaned up, but that's for tomorrow.

```
PS C:\dev\meson\test cases\common\13 pch\build> ninja -v c/prog.exe        
[1/12] "cl" "-Ic\libstlib_squared.a.p" "-Ic" "-I..\c" "-I..\c\pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/Ycstatic_squared_pch.h" "/Fpc\libstlib_squared.a.p\static_squared_pch.pch" "/Foc\libstlib_squared.a.p\static_squared_pch.obj" "/MDd" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/FS" "/FdC:\dev\meson\test cases\common\13 pch\build\c\libstlib_squared.pdb" "/c" c/libstlib_squared.a.p/meson_pch-c.c
[2/12] "cl" "-Ic\shared.dll.p" "-Ic" "-I..\c" "-I..\c\pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/Ycshared_pch.h" "/Fpc\shared.dll.p\shared_pch.pch" "/Foc\shared.dll.p\shared_pch.obj" "/MDd" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "-DBUILDING_DLL" "/FS" "/FdC:\dev\meson\test cases\common\13 pch\build\c\shared.pdb" "/c" c/shared.dll.p/meson_pch-c.c
[3/12] "cl" "-Ic\libstatic.a.p" "-Ic" "-I..\c" "-I..\c\pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/Ycstatic_pch.h" "/Fpc\libstatic.a.p\static_pch.pch" "/Foc\libstatic.a.p\static_pch.obj" "/MDd" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" /FS" "/FdC:\dev\meson\test cases\common\13 pch\build\c\libstatic.pdb" "/c" c/libstatic.a.p/meson_pch-c.c
[4/12] "cl" "-Ic\prog.exe.p" "-Ic" "-I..\c" "-I..\c\pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/Ycprog.h" "/Fpc\prog.exe.pprog.pch" "/Foc\prog.exe.p\prog.obj" "/MDd" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/FS" "/FdC:\devmeson\test cases\common\13 pch\build\c\prog.pdb" "/c" c/prog.exe.p/meson_pch-c.c
[5/12] "cl" "-Ic\shared.dll.p" "-Ic" "-I..\c" "/MDd" "/FIshared_pch.h" "/Yushared_pch.h" "/Fpc\shared.dll.p\shared_pch.pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "-DBUILDING_DLL" "/FS" "/FdC:\dev\meson\test cases\common\13 pch\buildcshared.pdb" /Foc/shared.dll.p/shared.c.obj "/c" ../c/shared.c       
[6/12] "cl" "-Ic\libstlib_squared.a.p" "-Ic" "-I..\c" "/MDd" "/FIstatic_squared_pch.h" "/Yustatic_squared_pch.h" "/Fpc\libstlib_squared.a.p\static_squared_pch.pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/FS" "/FdC:\dev\meson\test cases\common\13 pch\build\c\libstlib_squared.pdb" /Foc/libstlib_squared.a.p/static_squared.c.obj "/c" ../c/static_squared.c
[7/12] "cl" "-Ic\libstatic.a.p" "-Ic" "-I..\c" "/MDd" "/FIstatic_pch.h" "/Yustatic_pch.h" "/Fpc\libstatic.a.p\static_pch.pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/FS" "/FdC:\dev\meson\test cases\common\13 pch\build\c\libstatic.pdb" /Foc/libstatic.a.p/static.c.obj "/c" ../c/static.c
[8/12] "cl" "-Ic\prog.exe.p" "-Ic" "-I..\c" "/MDd" "/FIprog.h" "/Yuprog.h" "/Fpc\prog.exe.p\prog.pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/FS" "/FdC:\dev\meson\test cases\common\13 pch\build\c\prog.pdb" /Foc/prog.exe.p/prog.c.obj "/c" ../c/prog.c
[9/12] "lib" "/NOLOGO" "c\libstatic.a.p\static_pch.obj" /MACHINE:x64 /OUT:c/libstatic.a c/libstlib_squared.a.p/static_squared.c.obj c/libstatic.a.p/static.c.obj
[10/12] "link"  /MACHINE:x64 /OUT:c/shared.dll c/shared.dll.p/shared.c.obj "/nologo" "/release" "/nologo" "/DEBUG" "/PDB:c\shared.pdb" "/DLL" "/IMPLIB:c\shared.lib" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib" "c\shared.dll.p\shared_pch.obj"
   Creating library c\shared.lib and object c\shared.exp
[11/12] "C:\python\python.exe" "C:\dev\meson\meson.py" "--internal" "symbolextractor" "C:\dev\meson\test cases\common\13 pch\build" c/shared.dll "c\shared.lib" c/shared.dll.p/shared.dll.symbols
[12/12] "link"  /MACHINE:x64 /OUT:c/prog.exe c/prog.exe.p/prog.c.obj "/nologo" "/release" "/nologo" "/DEBUG" "/PDB:c\prog.pdb" "c\libstatic.a" "c\shared.lib" "/SUBSYSTEM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib" "c\prog.exe.p\prog.obj"
FAILED: c/prog.exe c/prog.pdb
"link"  /MACHINE:x64 /OUT:c/prog.exe c/prog.exe.p/prog.c.obj "/nologo" "/release" "/nologo" "/DEBUG" "/PDB:c\prog.pdb" "c\libstatic.a" "c\shared.lib" "/SUBSYST
EM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib" "c\prog.ex
e.p\prog.obj"
libstatic.a(static_squared.c.obj) : error LNK2011: precompiled object not linked in; image may not run
  Hint on symbols that are defined and could potentially match:
    __@@_PchSym_@00@UwveUnvhlmUgvhgLxzhvhUxlnnlmUBDLkxsUyfrowUxUkiltOvcvOkUkiltOlyq@4B2008FD98C1DD4
    __@@_PchSym_@00@UwveUnvhlmUgvhgLxzhvhUxlnnlmUBDLkxsUyfrowUxUoryhgzgrxOzOkUhgzgrxPkxsOlyq@4B2008FD98C1DD4
c\prog.exe : fatal error LNK1120: 1 unresolved externals
ninja: build stopped: subcommand failed.
```

The problem is
```
[9/12] "lib" "/NOLOGO" "c\libstatic.a.p\static_pch.obj" /MACHINE:x64 /OUT:c/libstatic.a c/libstlib_squared.a.p/static_squared.c.obj c/libstatic.a.p/static.c.obj
```
which is omitting c\libstlib_squared.a.p\static_squared_pch.obj, generated in [1/12].

Looks like CI in my own repo won't run windows tests? I'll open a draft PR to show the problem...

Regards,

Andres

**Comment by anarazel** [comment]
> For some reason the example needs link_whole, while postgres's case doesn't need link_whole. Assume somehow postgres's link_with gets transformed to a link_whole.

Yep:
```
                elif t.is_internal():
                    # When we're a static library and we link_with to an
                    # internal/convenience library, promote to link_whole.
                    return self.link_whole(t)
...

    def is_internal(self) -> bool:
        return not self.need_install

```

If I change stlib_squared to be install: false, and stlib to use link_with, the problem still triggers.

**Comment by anarazel** [comment]
Failure: https://dev.azure.com/jussi0947/jussi/_build/results?buildId=15020&view=logs&j=b4155ab4-dd6f-54ec-e75b-1188463f253f&t=bb87da09-8431-59d5-0b82-d398c3c1d8ee&l=1125

Given how MSVC's linking of PCH objects works, I don't think c94c49208 can fly that way. Either the pch object need to be returned by Target.extract_all_objects_recurse() or the "subsumed" convenience static libraries need to be tracked in a new list in BuildTarget. .oO(link_swalled_targets) and the backends learn to emit references to those.

**Comment by eli-schwartz** [comment]
> Assume somehow postgres's link_with gets transformed to a link_whole.

Yep, that should happen when you have an internal convenience library (used as an object library, is not installed) which is then added via link_with to another library that is resolved as a static library *and* is installed. In such a case, it makes no sense to add one static library as a command line argument to `ar`, and we cannot add its `-l` flag to the dependency tree via pkg-config files like we would with two installed static libraries (one of which depends on the other), so we pass its objects instead.

**Comment by dcbaker** [comment]
The objects have Reference to the target they’re extracted from, we use that in the backend already. I guess we need to do something in the backend with that information though with pch

**Comment by anarazel** [comment]
Building non-trivial things with meson HEAD on windows w/ msvc has been broken for three weeks... ISTM that #10699 ought to be reverted for now?

**Comment by xclaesse** [comment]
> ISTM that #10699 ought to be reverted for now?

well, it fixes a bug and creates another one... pick your poison...

Am I missing something or the fix is simply to add `self.pch.update(t.get_pch())` just after `self.objects += t.extract_all_objects_recurse()`?

**Comment by xclaesse** [comment]
@anarazel a first step to help us would be to open a PR with a test case that shows the issue.

**Comment by anarazel** [comment]
> @anarazel a first step to help us would be to open a PR with a test case that shows the issue.

I did: https://github.com/mesonbuild/meson/pull/10746

**Comment by dcbaker** [comment]
I have a windows machine again, I was planning to figure this out today, but then the first school flu struck

**Comment by eli-schwartz** [comment]
We urgently need this to be figured out and fixed, because 0.64.0 is going to be released soon.

**Comment by eli-schwartz** [comment]
Yes, I was already intending to revert the broken set of patchsets. I was just hoping against hope that I wouldn't have to because a fix would sneak in before final release. :D 

Unfortunately, @dcbaker has been unable to devote time to this due to various alternative requirements on his time, and I don't think he's going to be back again until after the final release.

No one else has volunteered to solve this either, so...reverting it we go.
